File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
public:
    int helper(int i, int j, int m, vector<vector<int>> &matrix, vector<vector<int>> &dp){
        if (j < 0 || j >= m)
        return 1e9;
        if (i == 0)
        return matrix[0][j];
        if (dp[i][j] != -1)
        return dp[i][j];

        int upr=matrix[i][j]+helper(i - 1, j + 1, m, matrix, dp);
        int up=matrix[i][j]+helper(i - 1, j, m, matrix, dp);
        int upl=matrix[i][j]+helper(i - 1, j - 1, m, matrix, dp);
        return dp[i][j]=min(upr,min(up,upl));
    }
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n=matrix.size();
        int m=matrix[0].size();
        vector<vector<int>>dp(n,vector<int>(m,-1));
        int mn=INT_MAX;
        for(int i=0;i<m;i++){
            int v=helper(n - 1, i, m, matrix, dp);
            mn=min(mn,v);
        }
        return mn;
    }
};

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    
    public int helper(int i,int j, int n, int m, int[][] matrix,int[][] dp){
        if(j<0 || j>=n){
            return Integer.MAX_VALUE;
        }
        if(i==n-1){
            return matrix[i][j];
        }

        if(dp[i][j]!=Integer.MAX_VALUE){
            return dp[i][j];
        }

        int dl=helper(i+1,j-1,n,m,matrix,dp);
        int db=helper(i+1,j,n,m,matrix,dp);
        int dr=helper(i+1,j+1,n,m,matrix,dp);

        return dp[i][j]=matrix[i][j]+Math.min(db,Math.min(dl,dr));
    }
    public int minFallingPathSum(int[][] matrix) {
        int n=matrix.length;
        int m=matrix[0].length;
        int ans=Integer.MAX_VALUE;
        int[][] dp=new int[n][n];
        for(int i=0;i<n;i++){
            Arrays.fill(dp[i],Integer.MAX_VALUE);
        }

        for(int j=0;j<n;j++){
            ans=Math.min(ans,helper(0,j,n,m,matrix,dp));
        }

        return ans;
    }
}